src/content/blog/rust-based-javascript-linters-lack-typed-linting-right-now/index.mdx

```diff
@@ -0,0 +1,151 @@
+That means linting performance will often be strictly worse than that of running `tsc` on your entire project.
+
+We're actively working on this in typescript-eslint.
+Our [Performance Troubleshooting](https://typescript-eslint.io/linting/troubleshooting/performance-troubleshooting) docs have some suggestions, and we're very hopeful that our [`EXPERIMENTAL_useProjectService` option](https://typescript-eslint.io/packages/parser/#experimental_useprojectservice) will land as stable in 2024.
+But even if that option works perfectly then type-checked linting will by design be orders of magnitude slower than traditional linting.
+
+Our experience has been that the benefits of type-checked linting almost always far outweigh the drawbacks of slower performance.
+Most of the time, when we've seen projects with slow type-checked linting, the root cause was either a misconfiguration of typescript-eslint or [slow TypeScript types](https://github.com/microsoft/TypeScript/wiki/Performance).
+
+## Rust-Based Linters and Type Checking
+
+**No Rust-based linter has integrated with TypeScript's type checking APIs yet**.
+That means no Rust-based linter is a full replacement for ESLint + typescript-eslint.
+
+I'm not saying you shouldn't use a Rust-based linter: if you don't want any of the type-checked lint rules, then sure, switching over is great.
+But I strongly recommend you look through at least the [recommended type-checked rules in typescript-eslint](https://typescript-eslint.io/rules/?=recommended-typeInformation) to understand what you're missing first.
+
+### Integrating Type-Checked Linting and Rust-Based Linters
+
+Right now, TypeScript is the only tool that can reliably provide type-checked linting for TypeScript code.
+TypeScript is written in TypeScript, so its type checking runs at JavaScript speed.
+
+In order to work with type checking, a Rust linter would have to either:
+
+- Take the performance hit of calling to TypeScript's JavaScript-speed type-checking APIs
+- Reimplement TypeScript's APIs in a native speed language
+- Speed Up TypeScript's APIs to native speed
+
+#### Option: Slowing Down to JavaScript Speed
+
+The performance hit option would likely slow the linters down to the point where they have no noticeable performance advantage compared to ESLint.
+👎
+
+That being said, if any native speed linter wants to do this, we in typescript-eslint would love to help.
+The [`@typescript-eslint/parser`](https://typescript-eslint.io/packages/parser) and [`@typescript-eslint/typescript-estree`](https://typescript-eslint.io/packages/typescript-estree) Node.js APIs are open source and as well documented as we've thought to write.
+We'd be happy to work with anybody who wants to use them, including spinning out standalone packages if that'd be useful.
+
+#### Option: Reimplementing TypeScript at Native Speed
```

Comment:
Oh, and I've been off-and-on wondering about whether a project could implement just the type _retrieval_ APIs - as opposed to _checking_. As in, it would only determine what types values are at time of declaration, and never perform assignability checks, evolving anys or arrays, etc.

Reply:
I tried porting type inference if that counts :-) https://twitter.com/boshen_c/status/1740397355051151730

Reply:
This is really interesting approach. You might be able to get something with a smaller subset only looking at variable, parameter and return type annotations and therefore skip subtyping logic. 

That is still quite complex, especially if you want this subset to cover generic calls and objects etc. and I assume in the long term you would want full type checking.